### ISSUE TITLE
Mod removed or privated from workshop unsubscribes and deletes from users....

### ISSUE THREAD

**Issue by Xetaxheb** [bluemagic123/tModLoader#2538]
### Version

Latest 1.4-stable

### OS

Windows

### Platform

Steam

### This bug affects

Gameplay as a Player

### Description

This is... a problem

For whatever reason the mods show up as "There was a problem accessing the item. Please try again." whether the author did something or whatever happened, but the underlying problem is that it is completely unacceptable both to have files removed without consent of end-users and obviously it could be a significant issue with breaking existing players/worlds.

**Comment by steviegt6** [comment]
This is a core functionality of the Steam workshop and technically an intended feature on their part. While an utterly massive inconvenience, there are perfectly acceptable reasons for privating or fully deleting a mod, such as:
* publishing private information,
* illegally publishing copyrighted content,
* publishing a mod under the wrong internal name,
* and so on.

**Comment by Xetaxheb** [comment]
Yes, and I'm not disputing the reason it went down, I have no idea what happened there although I contacted the mod author and **edit: they have no idea why it has disappeared from the workshop.** ~~previous: while it wasn't particularly fruitful of a conversation; from what they said it didn't seem like they had any idea it disappeared let alone why. (it's not something that would have been copyrighted and the internal was fine etc; maybe they were just deceiving/dismissing me? but it seems unlikely)~~

At any rate arguing why a mod goes off the workshop is all fine and well but not the point of my issue. Having a file taken off my computer without my permission or initiation is completely unacceptable, on top of the obvious problem with broken playthroughs.  

I'm not saying there needs to be a full takeover like nexusmods or curseforge (archiving and making available against the authors wishes), but somehow preventing end-users from having the rug pulled out from under them for any reason needs to happen. It's completely fine if something isn't available to download from the workshop and isn't condoned to redistribute, but the workshop and authors have no right to tell me what I can or can't keep on my computer.

[Additionally, it is literally against the steam subscriber agreement and therefore the steam workshop agreement 6.B.](https://store.steampowered.com/subscriber_agreement/#6)

Even if it's as simple as some kind of tmodloader toggle (off by default because disk space? but tmods aren't usually very big so idk if that should be a concern) for "backup workshop mods on update" and a "delete backups too?" on in-game trashcan. Ideally(?) this would do something like make a copy of every workshop mod with version appended from description (not modsources or loose mods for obvious reasons).
I get how it seems at first glance but this is one of the problems that really needs to be dealt with for the "luxury" of moving to steam workshop. The solution I'm suggesting would also allow users some wiggle room when a mod updates and breaks compatibility or preview version.

**Comment by JavidPack** [comment]
You can certainly copy the file over from the workshop folder to your mods folder, but reorganizing how mods are managed at this point might be a much bigger endeavor than we want to pursue. This is an interesting problem though. On one hand, removing potentially dangerous mods via moderation actions and having those automatically be unsubscribed, as it is now, would be good for the majority, but as you've mentioned it might be nice to be able to avoid that.

I can't promise we'll do anything about this anytime soon, but we'll keep this corner case in mind. The modpacks feature, when it is working, might just fix this for you, since I think it copies the current version of the mod to the modpack folder

**Comment by steviegt6** [comment]
I have my concerns with a lot of things regarding attempting to implement this behavior.

While I have a strong distaste for programs deleting my files for really any reason without by direct consent as well, I believe this is general Steam Workshop behavior, and this may put users at risk of using things like malicious or hijacked mods. Copying (backing up) mods also brings up an issue with the usage of additional storage space.

**Comment by Xetaxheb** [comment]
**TL;DR** **I think there's a simple and reasonable solution, file size isn't _that_ big and can be optional, and malicious mods a non-issue**
-
Mods aren't very big, and it can be optional.
**The most straightforward option to me would be last+current version backups.** 
1. Mod updates in /steamlibrary/workshop/
2. New version detected (already in place)
3. Existing file in /OldBackups/ is deleted
4. Existing file in /CurrentBackups/ moved to /OldBackups/
5. New file copied from /workshop/ to /CurrentBackups/

If the purpose was circumventing steamworkshop control, it has the obvious flaw of a mod updating multiple times between tmodloader launches missing files/backups... but I think this would serve the two objectives of 
- Mod got removed from store, my playthrough is screwed (_currentbackups has the fix_)
- Mod got updated and broke my playthrough (_oldbackups has the fix_)

quite well.

File size is a non-issue these days, this would at most increase mod content by 3x, which in the context of terraria goes from like 5GB with the biggest 300 mods installed to 15GB? More realistically for most people 2GB max to 6GB? (your numbers may vary, music packs, etc). Plus, optional.

I'm not sure "bad mods" likely enough to influence direction, and mod files _existing_ isn't going to hurt the computer, removing a mod off workshop WOULD unsubscribe and therefore disable the mod. 
**I'm not suggesting anything that doesn't involve a user manually and deliberately copying the backup back into /mods/ themselves.** This would also.. _basically solve a hijacked problem_ giving users access to a clean version of the mod.

Thank you for coming to my TED talk. 🥶 Sorry if parts seem curt, just trimmed from 5k. 

edit:
> I believe this is general Steam Workshop behavior

You're right, it is, but steam workshop is a platform with numerous logistical and inconvenient problems. It is arguably a bad distribution platform (albeit convenient and free? for tmodloader). I'm not saying something like curseforge or nexusmods don't have their share of issues, but they would be _better_ examples of competent distribution. Steam workshop is incredibly barebones in its features and requires developers to make any workarounds for systems they want in place. Like the 2022.x folders, instead of proper version compatibility and multiple file hosting. Why do I have two copies of MagicStorage.tmod in steamlibrary? Because the workshop is poorly designed 😂

**Comment by Jofairden** [comment]
> Additionally, it is literally against the steam subscriber agreement and therefore the steam workshop agreement 6.B.

You are right, it also says
> Specific Workshop-Enabled Apps or Workshop web pages may contain special terms (“App-Specific Terms”) that supplement or change the terms set out in this Section

We don't have that, but could make it. It's another option rather than changing the behaviour.
I think the way this was made is that it 'syncs' with the workshop, so when it gets removed from the workshop it gets flagged for deletion. I do agree there could be better handling to that, for example a warning that it can no longer update or as mentioned a configuration toggle that can inverse this behaviour.

I personally agree it should be changed. The legal document states very clearly once it gets distributed all the rights in the document can be claimed by the receiver
> Workshop Contributions are Subscriptions, and therefore you agree that any Subscriber receiving distribution of your Workshop Contribution will have the same rights to use your Workshop Contribution (and will be subject to the same restrictions) as are set out in this Agreement for any other Subscriptions.

My personal take is that a warning is a best bet, showing that it can't update anymore. I'm not that keen on the 'what if it's malicious code' argument. That has always been a risk. We have a custom EULA users of TML inherently agree on when playing TML through Steam that covers us for that. Obviously, we will remove any mod that is proven to be malicious.

**Comment by Jofairden** [comment]
We've discussed this internally and it's default behavior from SW, we simply have a folder that syncs with SW and the client simply follows that. If SW sync is deleting the file, that's on SW's end and not TML. We can provide an option potentially to restore a backup, but I don't think it's high up on the list atm.

**Comment by Chicken-Bones** [comment]
In short, you'd need to raise complaint with Steam about files being deleted off your computer if you want to pursue this further.
